File: typescript/src/native.ts
import { spawn, spawnSync } from "node:child_process";
import { isAbsolute } from "node:path";

import {
  boundedDiagnostic,
  resolveBinary,
  safeBootstrapEnvironment,
  ungatedEnvironment,
  validatedApiKey,
  validateProfileName,
  type LaunchDescriptor,
} from "./descriptor.js";
import { classifyRefusal, entitlementPreflight } from "./entitlement.js";
import {
  ChildProcessError,
  DeadlineExceededError,
  DescriptorError,
  EntitlementError,
  NativeRefusalError,
  OutputLimitError,
  ProcessCancelledError,
  ProtocolContractError,
} from "./errors.js";

const OPERATOR_OPERATIONS = new Set([
  "feedback",
  "memory_lifecycle",
  "memory_import",
  "address_maintenance",
  "maintenance",
  "ontology",
  "doctor",
]);
const PROFILE_KEYS = [
  "durability",
  "journal",
  "name",
  "principal_id",
  "root",
  "version",
  "workspace_id",
] as const;

export interface Profile {
  readonly version: 1;
  readonly name: string;
  readonly root: string;
  readonly workspace_id: string;
  readonly principal_id: string;
  readonly journal: string;
  readonly durability: string;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function parseJson(text: string | Buffer, label: string): unknown {
  try {
    return JSON.parse(Buffer.isBuffer(text) ? text.toString("utf8") : text);
  } catch (error) {
    throw new ProtocolContractError(`${label} did not return one JSON value`, { cause: error });
  }
}

export function parseProfile(value: unknown): Profile {
  if (!isRecord(value) || JSON.stringify(Object.keys(value).sort()) !== JSON.stringify(PROFILE_KEYS)) {
    throw new DescriptorError("profile differs from the closed v1 shape");
  }
  if (value.version !== 1) throw new DescriptorError("profile version must be exactly 1");
  const name = validateProfileName(value.name);
  const fields = ["root", "workspace_id", "principal_id", "journal", "durability"] as const;
  for (const field of fields) {
    if (typeof value[field] !== "string" || value[field].length === 0) {
      throw new DescriptorError(`profile ${field} is invalid`);
    }
  }
  if (!isAbsolute(value.root as string)) throw new DescriptorError("profile root must be absolute");
  return Object.freeze({
    version: 1,
    name,
    root: value.root as string,
    workspace_id: value.workspace_id as string,
    principal_id: value.principal_id as string,
    journal: value.journal as string,
    durability: value.durability as string,
  });
}

export function loadProfile(binary: string, name: string, timeoutMs = 10_000): Profile {
  const command = resolveBinary(binary);
  validateProfileName(name);
  const completed = spawnSync(command, ["profile", "show", name], {
    encoding: "utf8",
    // `profile show` is not a gated command.
    env: ungatedEnvironment(),
    timeout: timeoutMs,
    maxBuffer: 64 * 1024,
    windowsHide: true,
  });
  if (completed.error) {
    throw new ChildProcessError("profile show could not complete", { cause: completed.error });
  }
  if (completed.status !== 0) throw new ChildProcessError(`profile show exited ${completed.status}`);
  const profile = parseProfile(parseJson(completed.stdout, "profile show"));
  if (profile.name !== name) throw new ProtocolContractError("profile show changed the name");
  return profile;
}

export function schema(binary: string, operation?: string, timeoutMs = 10_000): string {
  const command = resolveBinary(binary);
  if (
    operation !== undefined &&
    operation !== "search" &&
    operation !== "remember" &&
    !OPERATOR_OPERATIONS.has(operation)
  ) {
    throw new RangeError(`unknown public operation ${JSON.stringify(operation)}`);
  }
  const completed = spawnSync(command, ["schema", ...(operation ? [operation] : [])], {
    encoding: "utf8",
    // `schema` is not a gated command.
    env: ungatedEnvironment(),
    timeout: timeoutMs,
    maxBuffer: 1024 * 1024,
    windowsHide: true,
  });
  if (completed.error) throw new ChildProcessError("schema could not complete", { cause: completed.error });
  if (completed.status !== 0) throw new ChildProcessError(`schema exited ${completed.status}`);
  return completed.stdout;
}

function sortJson(value: unknown, active = new Set<object>()): unknown {
  if (value === null || typeof value === "string" || typeof value === "boolean") return value;
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new TypeError("JSON numbers must be finite");
    return value;
  }
  if (typeof value !== "object") throw new TypeError("native arguments contain a non-JSON value");
  if (active.has(value)) throw new TypeError("native arguments contain a JSON cycle");
  active.add(value);
  try {
    if (Array.isArray(value)) return value.map((item) => sortJson(item, active));
    const prototype = Object.getPrototypeOf(value);
    if (prototype !== Object.prototype && prototype !== null) {
      throw new TypeError("native arguments contain a non-plain JSON object");
    }
    return Object.fromEntries(
      Object.keys(value)
        .sort()
        .map((key) => [key, sortJson((value as Record<string, unknown>)[key], active)]),
    );
  } finally {
    active.delete(value);
  }
}

function encodePayload(arguments_: Record<string, unknown>): Buffer {
  let text: string | undefined;
  try {
    text = JSON.stringify(sortJson(arguments_));
  } catch (error) {
    throw new ProtocolContractError("native arguments are not JSON serializable", { cause: error });
  }
  if (text === undefined) throw new ProtocolContractError("native arguments are not JSON serializable");
  return Buffer.from(text, "utf8");
}

interface AttemptResult {
  readonly code: number | null;
  readonly stdout: Buffer;
  readonly stderr: Buffer;
}

interface AttemptOptions {
  readonly timeoutMs: number;
  readonly signal?: AbortSignal;
  readonly stdoutLimit: number;
  readonly stderrLimit: number;
  readonly apiKey?: string;
}

function runAttempt(
  command: string,
  args: readonly string[],
  payload: Buffer,
  options: AttemptOptions,
): Promise<AttemptResult> {
  return new Promise((resolve, reject) => {
    if (options.signal?.aborted) {
      reject(new ProcessCancelledError("native call was cancelled before launch"));
      return;
    }
    const child = spawn(command, args, {
      // `call` IS gated, so this is the one native spawn that gets a key.
      env: safeBootstrapEnvironment(
        options.apiKey === undefined ? {} : { apiKey: options.apiKey },
      ),
      shell: false,
      stdio: ["pipe", "pipe", "pipe"],
      windowsHide: true,
    });
    const stdout: Buffer[] = [];
    const stderr: Buffer[] = [];
    let stdoutBytes = 0;
    let stderrBytes = 0;
    let terminalError: Error | undefined;
    let settled = false;
    let forceKill: NodeJS.Timeout | undefined;

    const terminate = (): void => {
      child.kill("SIGTERM");
      forceKill = setTimeout(() => child.kill("SIGKILL"), 1_000);
      forceKill.unref();
    };
    const fail = (error: Error): void => {
      if (terminalError) return;
      terminalError = error;
      terminate();
    };
    child.stdout.on("data", (raw: Buffer | string) => {
      const chunk = Buffer.from(raw);
      stdoutBytes += chunk.length;
      if (stdoutBytes > options.stdoutLimit) {
        fail(new OutputLimitError(`native stdout exceeded ${options.stdoutLimit} bytes`));
      } else {
        stdout.push(chunk);
      }
    });
    child.stderr.on("data", (raw: Buffer | string) => {
      const chunk = Buffer.from(raw);
      stderrBytes += chunk.length;
      if (stderrBytes > options.stderrLimit) {
        fail(new OutputLimitError(`native stderr exceeded ${options.stderrLimit} bytes`));
      } else {
        stderr.push(chunk);
      }
    });
    const timeout = setTimeout(() => {
      fail(new DeadlineExceededError("native call timed out after send; outcome may be uncertain"));
    }, options.timeoutMs);
    timeout.unref();
    const onAbort = (): void => fail(new ProcessCancelledError("native call was cancelled"));
    options.signal?.addEventListener("abort", onAbort, { once: true });

    const cleanup = (): void => {
      clearTimeout(timeout);
      if (forceKill) clearTimeout(forceKill);
      options.signal?.removeEventListener("abort", onAbort);
    };
    child.once("error", (error) => {
      if (settled) return;
      settled = true;
      cleanup();
      reject(new ChildProcessError("native child could not start", { cause: error }));
    });
    child.once("close", (code) => {
      if (settled) return;
      settled = true;
      cleanup();
      if (terminalError) {
        reject(terminalError);
      } else {
        resolve({ code, stdout: Buffer.concat(stdout), stderr: Buffer.concat(stderr) });
      }
    });
    child.stdin.on("error", () => undefined);
    child.stdin.end(payload);
  });
}

export interface NativeCallOptions {
  readonly signal?: AbortSignal;
}

class NativeCaller {
  readonly descriptor: LaunchDescriptor;
  readonly timeoutMs: number;
  readonly attempts: number;
  readonly stdoutLimit: number;
  readonly stderrLimit: number;
  /** Validated for TRANSPORT at construction; never rendered by any repr. */
  readonly #apiKey: string | undefined;

  constructor(
    descriptor: LaunchDescriptor,
    options: {
      timeoutMs?: number;
      attempts: number;
      stdoutLimit?: number;
      stderrLimit?: number;
      apiKey?: string;
    },
  ) {
    this.descriptor = descriptor;
    this.timeoutMs = options.timeoutMs ?? 30_000;
    this.attempts = options.attempts;
    this.stdoutLimit = options.stdoutLimit ?? 8 * 1024 * 1024;
    this.stderrLimit = options.stderrLimit ?? 64 * 1024;
    this.#apiKey = validatedApiKey(options.apiKey);
  }

  protected get apiKey(): string | undefined {
    return this.#apiKey;
  }

  async callNative(
    operation: string,
    arguments_: Record<string, unknown>,
    options: NativeCallOptions = {},
  ): Promise<unknown> {
    // Before the attempt loop, so a completely unconfigured caller never spawns
    // the engine at all. Fails open on an engine with no gate.
    const gate = entitlementPreflight(
      this.descriptor.command,
      this.#apiKey === undefined ? {} : { apiKey: this.#apiKey },
    );
    const payload = encodePayload(arguments_);
    const deadline = performance.now() + this.timeoutMs;
    let lastFailure: Error | undefined;
    for (let attempt = 0; attempt < this.attempts; attempt += 1) {
      const remaining = deadline - performance.now();
      if (remaining <= 0) break;
      const attemptsLeft = this.attempts - attempt;
      try {
        const result = await runAttempt(
          this.descriptor.command,
          ["call", "--profile", this.descriptor.args[2], operation],
          payload,
          {
            timeoutMs: remaining / attemptsLeft,
            stdoutLimit: this.stdoutLimit,
            stderrLimit: this.stderrLimit,
            ...(this.#apiKey === undefined ? {} : { apiKey: this.#apiKey }),
            ...(options.signal === undefined ? {} : { signal: options.signal }),
          },
        );
        if (result.code !== 0) {
          // An entitlement refusal is deterministic: the second spawn refuses
          // identically, and reporting it as "native call crashed" would call a
          // deliberate refusal a crash. Classified before the JSON parse
          // because the engine writes nothing to stdout when it refuses.
          const reason = classifyRefusal(result.stderr, result.code);
          if (reason !== null) {
            const keyFile = gate.keyFile;
            throw new EntitlementError(reason, {
              diagnostic: boundedDiagnostic(result.stderr),
              ...(keyFile === undefined ? {} : { keyFile }),
            });
          }
        }
        let parsed: unknown;
        try {
          parsed = JSON.parse(result.stdout.toString("utf8"));
        } catch (error) {
          if (result.code === 0) {
            throw new ProtocolContractError("native child returned non-JSON on success", {
              cause: error,
            });
          }
          lastFailure = new ChildProcessError(
            `native child exited ${result.code ?? "without status"} before a JSON response`,
          );
          continue;
        }
        if (result.code !== 0) throw new NativeRefusalError(operation, parsed);
        return parsed;
      } catch (error) {
        if (
          // Declared, not load-bearing today: EntitlementError extends
          // IntegrationError rather than ChildProcessError, so removing this
          // line changes no outcome -- measured, by removing it and watching
          // the non-retry test stay green. It is here so a future change that
          // reparents EntitlementError cannot silently make a deterministic
          // refusal retryable. What actually proves the property is the spawn
          // counter in the non-retry test, with a control that reads 2.
          error instanceof EntitlementError ||
          error instanceof NativeRefusalError ||
          error instanceof OutputLimitError ||
          error instanceof ProcessCancelledError ||
          error instanceof ProtocolContractError
        ) {
          throw error;
        }
        if (error instanceof DeadlineExceededError || error instanceof ChildProcessError) {
          lastFailure = error;
          continue;
        }
        throw error;
      }
    }
    if (lastFailure instanceof DeadlineExceededError) {
      throw new DeadlineExceededError(
        "native call exhausted its original deadline after one bounded retry",
        { cause: lastFailure },
      );
    }
    throw new ChildProcessError(
      `native call crashed before a response after ${this.attempts - 1} bounded retries`,
      { cause: lastFailure },
    );
  }
}

export class Controller extends NativeCaller {
  constructor(
    descriptor: LaunchDescriptor,
    options: {
      timeoutMs?: number;
      stdoutLimit?: number;
      stderrLimit?: number;
      apiKey?: string;
    } = {},
  ) {
    super(descriptor, { ...options, attempts: 2 });
  }

  searchRaw(arguments_: Record<string, unknown>, options?: NativeCallOptions): Promise<unknown> {
    return this.callNative("search", arguments_, options);
  }

  rememberRaw(arguments_: Record<string, unknown>, options?: NativeCallOptions): Promise<unknown> {
    return this.callNative("remember", arguments_, options);
  }
}

export class Operator extends NativeCaller {
  constructor(
    descriptor: LaunchDescriptor,
    options: { timeoutMs?: number; apiKey?: string } = {},
  ) {
    super(descriptor, { ...options, attempts: 1 });
  }

  call(
    operation: string,
    arguments_: Record<string, unknown>,
    options?: NativeCallOptions,
  ): Promise<unknown> {
    if (!OPERATOR_OPERATIONS.has(operation)) {
      throw new RangeError(`${JSON.stringify(operation)} is not in the operator namespace`);
    }
    return this.callNative(operation, arguments_, options);
  }
}

File: python/src/kaleidoscope_memory/native.py
"""Direct native controller and explicit operator namespace.

These objects are application/controller APIs. They are never exposed to a
model as tools; model-facing use stays on the two-tool MCP surface.
"""

from __future__ import annotations

import asyncio
import json
import math
import subprocess
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Mapping

from .descriptor import (
    LaunchDescriptor,
    _bounded_diagnostic,
    _canonical_executable,
    _Secret,
    _safe_process_environment,
    _ungated_environment,
    executable_sha256,
    hold_api_key,
    reveal_api_key,
    validate_profile_name,
)
from .entitlement import classify_refusal, entitlement_preflight
from .errors import (
    ChildProcessError,
    DeadlineExceededError,
    DescriptorError,
    EntitlementError,
    NativeRefusalError,
    OutputLimitError,
    ProtocolError,
)

_AGENT_OPERATIONS = frozenset({"search", "remember"})
_OPERATOR_OPERATIONS = frozenset(
    {
        "feedback",
        "memory_lifecycle",
        "memory_import",
        "address_maintenance",
        "maintenance",
        "ontology",
        "doctor",
    }
)
_PROFILE_KEYS = frozenset(
    {
        "version",
        "name",
        "root",
        "workspace_id",
        "principal_id",
        "journal",
        "durability",
    }
)


@dataclass(frozen=True, slots=True)
class Profile:
    version: int
    name: str
    root: str
    workspace_id: str
    principal_id: str
    journal: str
    durability: str

    @classmethod
    def from_mapping(cls, value: Mapping[str, Any]) -> "Profile":
        if set(value) != _PROFILE_KEYS or type(value.get("version")) is not int or value["version"] != 1:
            raise DescriptorError("profile differs from the closed v1 shape")
        name = validate_profile_name(value["name"])
        strings = ("root", "workspace_id", "principal_id", "journal", "durability")
        if any(not isinstance(value.get(key), str) or not value[key] for key in strings):
            raise DescriptorError("profile contains an invalid identity field")
        root = Path(value["root"])
        if not root.is_absolute():
            raise DescriptorError("profile root must be absolute")
        return cls(
            version=1,
            name=name,
            root=str(root),
            workspace_id=value["workspace_id"],
            principal_id=value["principal_id"],
            journal=value["journal"],
            durability=value["durability"],
        )


def resolve_binary(candidate: str | Path | None = None, *, expected_sha256: str | None = None) -> str:
    if candidate is None:
        from .distribution import locate_engine

        candidate = locate_engine().path
    command = _canonical_executable(str(candidate))
    if expected_sha256 is not None and executable_sha256(command) != expected_sha256.lower():
        raise DescriptorError("Kaleidoscope executable SHA-256 does not match the caller's pin")
    return command


def resolve_manager(candidate: str | Path | None = None, *, expected_sha256: str | None = None) -> str:
    if candidate is None:
        from .distribution import locate_manager

        candidate = locate_manager().path
    command = _canonical_executable(str(candidate))
    if expected_sha256 is not None and executable_sha256(command) != expected_sha256.lower():
        raise DescriptorError("Kaleidoscope manager SHA-256 does not match the caller's pin")
    return command


def mcp_stdio_config(descriptor: LaunchDescriptor) -> dict[str, object]:
    """SDK-neutral config; omitted env means the SDK's safe bootstrap allowlist."""

    return {"command": descriptor.command, "args": list(descriptor.args)}


def load_profile(binary: str | Path, name: str, *, timeout_seconds: float = 10.0) -> Profile:
    command = resolve_binary(binary)
    validate_profile_name(name)
    value = _run_manager_json(command, ["profile", "show", name], timeout_seconds)
    if not isinstance(value, dict):
        raise ProtocolError("profile show must return one JSON object")
    profile = Profile.from_mapping(value)
    if profile.name != name:
        raise ProtocolError("profile show changed the requested name")
    return profile


def schema(binary: str | Path, operation: str | None = None, *, timeout_seconds: float = 10.0) -> str:
    command = resolve_binary(binary)
    if operation is not None and operation not in _AGENT_OPERATIONS | _OPERATOR_OPERATIONS:
        raise ValueError(f"unknown public operation {operation!r}")
    argv = [command, "schema", *([operation] if operation is not None else [])]
    try:
        completed = subprocess.run(
            argv,
            check=False,
            capture_output=True,
            # `schema` is not in the engine's gated command list.
            env=_ungated_environment(),
            timeout=timeout_seconds,
        )
    except subprocess.TimeoutExpired as exc:
        raise DeadlineExceededError("schema deadline elapsed") from exc
    except OSError as exc:
        raise ChildProcessError("schema process could not start") from exc
    if completed.returncode != 0:
        raise ChildProcessError(f"schema exited {completed.returncode}")
    try:
        return completed.stdout.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise ProtocolError("schema output was not UTF-8") from exc


def _run_manager_json(command: str, args: list[str], timeout_seconds: float) -> Any:
    try:
        completed = subprocess.run(
            [command, *args],
            check=False,
            capture_output=True,
            # `profile show`/`profile list` are not gated commands.
            env=_ungated_environment(),
            timeout=timeout_seconds,
        )
    except subprocess.TimeoutExpired as exc:
        raise DeadlineExceededError("manager deadline elapsed") from exc
    except OSError as exc:
        raise ChildProcessError("manager process could not start") from exc
    if completed.returncode != 0:
        raise ChildProcessError(f"manager exited {completed.returncode}")
    try:
        return json.loads(completed.stdout)
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        raise ProtocolError("manager output was not one JSON value") from exc


async def _read_bounded(stream: asyncio.StreamReader, limit: int, label: str) -> bytes:
    chunks: list[bytes] = []
    total = 0
    while chunk := await stream.read(64 * 1024):
        total += len(chunk)
        if total > limit:
            raise OutputLimitError(f"native {label} exceeded {limit} bytes")
        chunks.append(chunk)
    return b"".join(chunks)


async def _communicate_bounded(
    process: asyncio.subprocess.Process,
    payload: bytes,
    *,
    stdout_limit: int,
    stderr_limit: int,
) -> tuple[bytes, bytes]:
    assert process.stdin is not None and process.stdout is not None and process.stderr is not None

    async def write() -> None:
        try:
            process.stdin.write(payload)
            await process.stdin.drain()
        except (BrokenPipeError, ConnectionResetError):
            pass
        finally:
            process.stdin.close()

    stdout_task = asyncio.create_task(_read_bounded(process.stdout, stdout_limit, "stdout"))
    stderr_task = asyncio.create_task(_read_bounded(process.stderr, stderr_limit, "stderr"))
    write_task = asyncio.create_task(write())
    try:
        stdout, stderr, _ = await asyncio.gather(stdout_task, stderr_task, write_task)
        await process.wait()
        return stdout, stderr
    except BaseException:
        for task in (stdout_task, stderr_task, write_task):
            task.cancel()
        raise


async def _terminate(process: asyncio.subprocess.Process) -> None:
    if process.returncode is not None:
        return
    process.terminate()
    try:
        await asyncio.wait_for(process.wait(), timeout=1.0)
    except TimeoutError:
        process.kill()
        await process.wait()


class _NativeCaller:
    def __init__(
        self,
        descriptor: LaunchDescriptor,
        *,
        timeout_seconds: float = 30.0,
        attempts: int,
        stdout_limit: int = 8 * 1024 * 1024,
        stderr_limit: int = 64 * 1024,
        api_key: str | None = None,
    ) -> None:
        self._descriptor = descriptor
        self._timeout_seconds = timeout_seconds
        self._attempts = attempts
        self._stdout_limit = stdout_limit
        self._stderr_limit = stderr_limit
        # Validated where the caller wrote it; held so no repr can render it.
        self._api_key: _Secret | None = hold_api_key(api_key)

    async def _call(self, operation: str, arguments: Mapping[str, Any]) -> Any:
        # `call` is a gated command. Refuse once, before the attempt loop, when
        # the engine enforces the gate and nothing is configured. Fails open.
        #
        # The returned status is quoted for the rest of this call rather than
        # re-probed on the error path. Asking twice let the two answers differ,
        # and a refusal naming a key file the user never configured is a refusal
        # spelled as the wrong answer.
        #
        # Revealed INLINE and never bound to a name: a `key = reveal_api_key(...)`
        # local here survives the whole retry loop, and every `await` in it is a
        # point at which an exception can be raised with this frame still on the
        # stack. Any instrument that captures frame locals then prints the
        # credential. Measured before this change with
        # `TracebackException(capture_locals=True)` against the real engine.
        gate = entitlement_preflight(
            self._descriptor.command, api_key=reveal_api_key(self._api_key)
        )
        try:
            _validate_json_value(arguments)
            payload = json.dumps(
                arguments,
                sort_keys=True,
                separators=(",", ":"),
                allow_nan=False,
            ).encode("utf-8")
        except (TypeError, ValueError, RecursionError) as exc:
            raise ProtocolError("native arguments are not a closed JSON value") from exc
        deadline = time.monotonic() + self._timeout_seconds
        last_failure: BaseException | None = None
        for attempt in range(self._attempts):
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                break
            attempts_left = self._attempts - attempt
            attempt_budget = remaining / attempts_left
            try:
                process = await asyncio.create_subprocess_exec(
                    self._descriptor.command,
                    "call",
                    "--profile",
                    self._descriptor.profile,
                    operation,
                    stdin=asyncio.subprocess.PIPE,
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.PIPE,
                    # `call` IS gated, so this is the one native spawn that gets
                    # the credential. Built inline for the same reason it is
                    # revealed inline above: the dict is a temporary on this
                    # frame's value stack, not a local anything can capture.
                    env=_safe_process_environment(
                        api_key=reveal_api_key(self._api_key)
                    ),
                )
            except OSError as exc:
                last_failure = ChildProcessError("native child could not start")
                last_failure.__cause__ = exc
                continue
            try:
                stdout, stderr = await asyncio.wait_for(
                    _communicate_bounded(
                        process,
                        payload,
                        stdout_limit=self._stdout_limit,
                        stderr_limit=self._stderr_limit,
                    ),
                    timeout=attempt_budget,
                )
            except asyncio.CancelledError:
                await _terminate(process)
                raise
            except TimeoutError as exc:
                await _terminate(process)
                last_failure = DeadlineExceededError(
                    "native call timed out after send; final outcome may be uncertain"
                )
                last_failure.__cause__ = exc
                continue
            except OutputLimitError:
                await _terminate(process)
                raise

            parsed: Any | None = None
            try:
                parsed = json.loads(stdout)
            except (json.JSONDecodeError, UnicodeDecodeError):
                pass
            if process.returncode != 0:
                reason = classify_refusal(stderr, process.returncode)
                if reason is not None:
                    # Deterministic: the second spawn refuses identically, so
                    # this raises rather than `continue`, joining
                    # NativeRefusalError and OutputLimitError in the
                    # non-retryable set. It is also a deliberate refusal, not a
                    # crash, so it is not a ChildProcessError.
                    raise EntitlementError(
                        reason,
                        diagnostic=_bounded_diagnostic(stderr),
                        key_file=gate.key_file,
                    )
                if parsed is not None:
                    raise NativeRefusalError(operation, parsed)
                last_failure = ChildProcessError(
                    f"native child exited {process.returncode} before a JSON response"
                )
                continue
            if parsed is None:
                raise ProtocolError("native child returned non-JSON on a successful exit")
            return parsed

        if isinstance(last_failure, DeadlineExceededError):
            raise DeadlineExceededError(
                "native call exhausted its original deadline after one bounded retry"
            ) from last_failure
        if last_failure is not None:
            raise ChildProcessError("native call crashed before a response after one bounded retry") from last_failure
        raise DeadlineExceededError("native call deadline elapsed before launch")


def _validate_json_value(value: Any, active: set[int] | None = None) -> None:
    if value is None or isinstance(value, (str, bool, int)):
        return
    if isinstance(value, float):
        if not math.isfinite(value):
            raise ValueError("JSON numbers must be finite")
        return
    if active is None:
        active = set()
    if isinstance(value, Mapping):
        if any(not isinstance(key, str) for key in value):
            raise TypeError("JSON object keys must be strings")
        identity = id(value)
        if identity in active:
            raise ValueError("JSON values cannot contain cycles")
        active.add(identity)
        try:
            for item in value.values():
                _validate_json_value(item, active)
        finally:
            active.remove(identity)
        return
    if isinstance(value, list):
        identity = id(value)
        if identity in active:
            raise ValueError("JSON values cannot contain cycles")
        active.add(identity)
        try:
            for item in value:
                _validate_json_value(item, active)
        finally:
            active.remove(identity)
        return
    raise TypeError(f"unsupported JSON value {type(value).__name__}")


class Controller(_NativeCaller):
    """Private native JSON access for application-owned acquisition/write paths."""

    def __init__(
        self,
        descriptor: LaunchDescriptor,
        *,
        timeout_seconds: float = 30.0,
        stdout_limit: int = 8 * 1024 * 1024,
        stderr_limit: int = 64 * 1024,
        api_key: str | None = None,
    ) -> None:
        super().__init__(
            descriptor,
            timeout_seconds=timeout_seconds,
            attempts=2,
            stdout_limit=stdout_limit,
            stderr_limit=stderr_limit,
            api_key=api_key,
        )

    async def search_raw(self, arguments: Mapping[str, Any]) -> Any:
        return await self._call("search", arguments)

    async def remember_raw(self, arguments: Mapping[str, Any]) -> Any:
        return await self._call("remember", arguments)


class Operator(_NativeCaller):
    """Explicit non-model namespace; operator calls are never advertised over MCP."""

    def __init__(
        self,
        descriptor: LaunchDescriptor,
        *,
        timeout_seconds: float = 30.0,
        api_key: str | None = None,
    ) -> None:
        # Operator semantics differ by verb, so staging does not auto-retry them.
        super().__init__(
            descriptor,
            timeout_seconds=timeout_seconds,
            attempts=1,
            api_key=api_key,
        )

    async def call(self, operation: str, arguments: Mapping[str, Any]) -> Any:
        if operation not in _OPERATOR_OPERATIONS:
            raise ValueError(f"operation {operation!r} is not in the operator namespace")
        return await self._call(operation, arguments)
